### PR TITLE
Refactor rmsprop `mean_square` and add unit tests for optimizers 

### PR DESCRIPTION
**Pull request by Ldpe2G**
- [x] 把 rmsprop 优化器 op 的 `mean_square` 从  temp buffer 改为 variable op ；
- [x] 给优化器op添加单元测试：
  - [x] rmsprop
  - [x] adam
  - [x] sgd
  - [x] adamw
  - [x] lazy_adam
  - [x] lars 

tensorflow 2.0 并没有包含 lars，lazyadam和 adamw，所以采用 与 numpy 实现对比的形式。

给 optimizers 添加单测的方式是，对一个随机初始化的 variable 做 n（目前是10）次更新迭代 ，然后看与 tensorflow 或者 numpy 的实现是否结果相近（误差限制在 1e-4）。

### CODE REVIEW

File: oneflow/core/job_rewriter/rmsprop_optm.cpp [hunk 1]
```diff
@@ -19,14 +19,38 @@ namespace oneflow {
 
 namespace {
 
+OperatorConf GenerateAdamHelperVariableOpConf(const VariableOp& op, const std::string& name,
```
Comment (liujuncheng) [inline oneflow/core/job_rewriter/rmsprop_optm.cpp:22]: 这个命名里面还包括`Adam`

File: oneflow/core/job_rewriter/rmsprop_optm.cpp [hunk 2]
```diff
@@ -19,14 +19,38 @@ namespace oneflow {
...
+                                              const float initial_value) {
+  OperatorConf helper_variable_op(op.op_conf());
+  helper_variable_op.set_name(op.op_name() + "-" + name);
+  helper_variable_op.mutable_variable_conf()->set_out("out");
+  InitializerConf constant_initializer;
+  constant_initializer.mutable_constant_conf()->set_value(initial_value);
+  *(helper_variable_op.mutable_variable_conf()->mutable_initializer()) = constant_initializer;
```
Comment (liujuncheng) [inline oneflow/core/job_rewriter/rmsprop_optm.cpp:29]: `helper_variable_op.set_scope_symbol_id(op.op_conf().scope_symbol_id());`
这样如何？

File: oneflow/core/kernel/rmsprop_model_update_kernel.cu [hunk 3]
```diff
@@ -22,14 +22,20 @@ namespace {
...
-    model[i] = model[i] - *learning_rate * model_diff_val / std::sqrt(mean_square[i] + epsilon);
+        (1 - decay_rate) * model_diff_val * model_diff_val + decay_rate * mean_square[i];
+    T denom_t;
+    if (centered) {
+      mean_gradient[i] = (1 - decay_rate) * model_diff_val + decay_rate * mean_gradient[i];
+      denom_t = mean_square[i] - mean_gradient[i] * mean_gradient[i];
+    } else {
+      denom_t = mean_square[i];
+    }
+    model[i] = model[i] - *learning_rate * model_diff_val / std::sqrt(denom_t + epsilon);
```
Comment (liujuncheng) [inline oneflow/core/kernel/rmsprop_model_update_kernel.cu:38]: 把 `/ std::sqrt(denom_t + epsilon)` 换成 ` * rsqrt(denom_t + epsilon)`  是否可行，性能应该会更好

File: oneflow/core/kernel/rmsprop_model_update_kernel.cu [hunk 4]
```diff
@@ -22,14 +22,20 @@ namespace {
...
   CUDA_1D_KERNEL_LOOP(i, n) {
     T model_diff_val = model_diff[i];
     mean_square[i] =
-        (1 - cur_decay_rate) * model_diff_val * model_diff_val + cur_decay_rate * mean_square[i];
-    model[i] = model[i] - *learning_rate * model_diff_val / std::sqrt(mean_square[i] + epsilon);
+        (1 - decay_rate) * model_diff_val * model_diff_val + decay_rate * mean_square[i];
+    T denom_t;
+    if (centered) {
+      mean_gradient[i] = (1 - decay_rate) * model_diff_val + decay_rate * mean_gradient[i];
+      denom_t = mean_square[i] - mean_gradient[i] * mean_gradient[i];
```
Comment (liujuncheng) [inline oneflow/core/kernel/rmsprop_model_update_kernel.cu:34]: 将计算好的 `mean_square[i]` 放到临时变量，可以避免多次从显存中加载

File: oneflow/core/kernel/rmsprop_model_update_kernel.cu [hunk 5]
```diff
@@ -22,14 +22,20 @@ namespace {
...
+                               T decay_rate, T epsilon, bool centered, T weight_decay,
+                               const T* model_diff, T* model, T* mean_square, T* mean_gradient) {
   CUDA_1D_KERNEL_LOOP(i, n) {
     T model_diff_val = model_diff[i];
     mean_square[i] =
-        (1 - cur_decay_rate) * model_diff_val * model_diff_val + cur_decay_rate * mean_square[i];
-    model[i] = model[i] - *learning_rate * model_diff_val / std::sqrt(mean_square[i] + epsilon);
+        (1 - decay_rate) * model_diff_val * model_diff_val + decay_rate * mean_square[i];
+    T denom_t;
+    if (centered) {
```
Comment (liujuncheng) [inline oneflow/core/kernel/rmsprop_model_update_kernel.cu:32]: 如果可以，将`centered`作为模板参数，可以消除这里if的开销

File: oneflow/python/test/ops/test_optimizers.py [hunk 6]
```diff
@@ -0,0 +1,513 @@
...
+    def testSGD() -> flow.typing.Numpy:
+        with flow.scope.placement(device_type, "0:0-0"):
+            x = flow.get_variable(
+                name="x",
+                shape=x_shape,
+                dtype=flow.float32,
+                initializer=flow.random_uniform_initializer(minval=0, maxval=100),
+                trainable=True,
+            )
+            loss = flow.math.reduce_mean(x)
```
Comment (liujuncheng) [inline oneflow/python/test/ops/test_optimizers.py:410]: 这里可以再引入个随机数和x做乘法计算，或者改成`loss = flow.math.reduce_mean(x * x)`应该也行，否则这里回传过去的梯度将一直是 `1 / x_shape.elem_cnt()`

File: oneflow/python/test/ops/test_optimizers.py [hunk 7]
```diff
@@ -0,0 +1,513 @@
...
+    arg_dict["train_iters"] = [10]
+    for arg in GenArgList(arg_dict):
+        compare_with_numpy_lars(*arg)
+
+
+def test_sgd(test_case):
+    arg_dict = OrderedDict()
+    arg_dict["device_type"] = ["cpu", "gpu"]
+    arg_dict["x_shape"] = [(10,)]
+    arg_dict["momentum"] = [0.9]
```
Comment (liujuncheng) [inline oneflow/python/test/ops/test_optimizers.py:509]: 加个momentum = 0.0吧

File: oneflow/core/kernel/rmsprop_model_update_kernel.cu (hunk 5 above)
Comment (Ldpe2G) [reply oneflow/core/kernel/rmsprop_model_update_kernel.cu]: 这里尝试了下，把centered放模板，但是没成功，这里改成，centered true 和 false 分别写两个函数，就在外面根据 centered 的值判断是调哪个 kernel，cpu 也是这样改，这样子 if 判断只会调用一次。

Comment (liujuncheng) [reply oneflow/core/kernel/rmsprop_model_update_kernel.cu]: > 这里尝试了下，把centered放模板，但是没成功，这里改成，centered true 和 false 分别写两个函数，就在外面根据 centered 的值判断是调哪个 kernel，cpu 也是这样改，这样子 if 判断只会调用一次。

可以，我今天改另外一个kernel时测试发现，这里这个if对性能可能基本没有影响

File: oneflow/core/kernel/rmsprop_model_update_kernel.cu [hunk 8]
```diff
@@ -21,15 +21,36 @@ namespace oneflow {
...
+    mean_square_val =
+        (1 - decay_rate) * model_diff_val * model_diff_val + decay_rate * mean_square_val;
+    mean_square[i] = mean_square_val;
+    T mean_gradient_val = mean_gradient[i];
+    mean_gradient_val = (1 - decay_rate) * model_diff_val + decay_rate * mean_gradient_val;
+    mean_gradient[i] = mean_gradient_val;
+    T denom_t = mean_square_val - mean_gradient_val * mean_gradient_val;
+    model[i] = model[i] - *learning_rate * model_diff_val * rsqrt(denom_t + epsilon);
+  }
+}
```
Comment (liujuncheng) [inline oneflow/core/kernel/rmsprop_model_update_kernel.cu:40]: 这里可以这样写
```cuda
template<typename T, bool centered>
__global__ void UpdateModelGpu(int64_t n, const int64_t* train_step, const float* learning_rate,
                               T decay_rate, T epsilon, T weight_decay, const T* model_diff,
                               T* model, T* mean_square, T* mean_gradient) {
  CUDA_1D_KERNEL_LOOP(i, n) {
    T model_diff_val = model_diff[i];
    T mean_square_val = mean_square[i];
    mean_square_val =
        (1 - decay_rate) * model_diff_val * model_diff_val + decay_rate * mean_square_val;
    mean_square[i] = mean_square_val;
    T denom_t;
    if (centered) {
      T mean_gradient_val = mean_gradient[i];
      mean_gradient_val = (1 - decay_rate) * model_diff_val + decay_rate * mean_gradient_val;
      mean_gradient[i] = mean_gradient_val;
      denom_t = mean_square_val - mean_gradient_val * mean_gradient_val;
    } else {
      denom_t = mean_square_val;
    }
    model[i] = model[i] - *learning_rate * model_diff_val * rsqrt(denom_t + epsilon);
  }
}
```
然后调用的地方
```c++
    if (centered) {
      UpdateModelGpu<T, true>
          <<<BlocksNum4ThreadsNum(n), kCudaThreadsNumPerBlock, 0, ctx->cuda_stream()>>>(
              n, train_step, learning_rate, decay_rate, epsilon, weight_decay, model_diff, model,
              mean_square, mean_gradient);
    } else {
      UpdateModelGpu<T, false>
          <<<BlocksNum4ThreadsNum(n), kCudaThreadsNumPerBlock, 0, ctx->cuda_stream()>>>(
              n, train_step, learning_rate, decay_rate, epsilon, weight_decay, model_diff, model,
              mean_square, mean_gradient);
    }
```
就可以了，可以减少重复代码，编译的时候其中一个分支会被直接优化掉

**Review by liujuncheng — APPROVED** [review]
